### PR TITLE
New Article: Create Your own Excel Using Perl!

### PR DESCRIPTION
**Pull request by x1mandi**
Hi,
Submitting the article about Excel::Writer::XLSX.
Tested, generated preview and proofread it, looks fine to me. 

Please let me know Your thoughts.

Kind regards,
Csaba

### CODE REVIEW

**Comment by thibaultduponchelle** [comment]
Great 👍 

(can you rebase your branch?)

**Comment by x1mandi** [comment]
> Great 👍
> 
> (can you rebase your branch?)

I think I did it, but not sure... I hope I did not mess up anything.

**Comment by thibaultduponchelle** [comment]
@oodler577 @openstrike May we ask you to help review this article? 🙏🏼

File: 2022/incoming/create-your-excel-using-perl.pod [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+	my $baking_sheet = $merry->add_worksheet('Xmas');
+	my $stats = $merry->add_worksheet('Everybody!');
+	$xmas->write( 'A1', 'Merry Christmas!' );
+	$merry->close();
+
+=end code
+
+I skipped to display the usual stuff from the beginning of the script, everyone should know by heart: C<use strict; use warnings;>, etc. 
+This will be the skeleton or the Christmas themed frame of our script.
+We can put everything else in between, depending on Santa's needs, a list of Christmas cookies from around Europe in hash format, is a good example, including some stats.
```
Comment (openstrike) [inline 2022/incoming/create-your-excel-using-perl.pod:44]: s/format, is/format is/

File: 2022/incoming/create-your-excel-using-perl.pod [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+		     
+	my $format = $workbook->add_format( %props );
+	#write welcome text with formatting to top left cell
+	$xmas->write( 'A1', 'Merry Christmas!', $format);
+	#Set the column width.
+	$xmas->set_column('A:A', 30);
+
+=end code
+
+You will need to pass the format object to Your write method as parameter:
```
Comment (openstrike) [inline 2022/incoming/create-your-excel-using-perl.pod:70]: s/as parameter/as a parameter/

File: 2022/incoming/create-your-excel-using-perl.pod [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+	#Just some snowflakes as border
+	my $format3 = $merry->add_format( border => 4,);
+
+=end code
+
+=head2 Writing data into cells
+
+There are two ways to do that: You can use either the row-column notation or the A1 notation as described in L<Cell notation|https://metacpan.org/pod/Excel::Writer::XLSX#Cell-notation>.
+First of all we need some data, let's compose our cookies list.
```
Comment (openstrike) [inline 2022/incoming/create-your-excel-using-perl.pod:102]: s/data, let's/data so let's/

File: 2022/incoming/create-your-excel-using-perl.pod [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+	  $baking_sheet->write( 1, $i, [ $cookies->{$_} ], $format3 );
+	  $i++;
+	}
+
+	#write some statistics about how many types of cookies we have per country
+	$stats->write( 'A1', [\@countries], $format2 );
+
+	#Reset the counter
+	$i = 0;
+	#count how many cookies we country wise.
```
Comment (openstrike) [inline 2022/incoming/create-your-excel-using-perl.pod:137]: "we have", perhaps?

File: 2022/incoming/create-your-excel-using-perl.pod [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        
+=end code
+
+Many things are happening here.
+
+First we get a list of countries sorted alphabetically.
+Second we are writing those countries into the first row, alongside some formats, defined above.
+Then comes the first usage of C<write()>, where we are using the A1 notation to start from the top left corner, then comes the reference to the C<@countries> array, followed by the format2.
+When we pass an array reference to C<write()>, it will write the elements of the array in the same row. It is equivalent to write_row().
+If You pass and array ref of array refs to C<write()>, it is the same if You would call C<write_col()>, it will write to the columns, as demonstrated in the lines 22 and 27.
```
Comment (openstrike) [inline 2022/incoming/create-your-excel-using-perl.pod:152]: s/pass and array/pass an array/
s/same if You would call/same as calling/

File: 2022/incoming/create-your-excel-using-perl.pod [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+For more on C<write()>, please check the L<write section|https://metacpan.org/pod/Excel::Writer::XLSX#write(-$row,-$column,-$token,-$format-)> on the module's page.
+
+=head2 Outro
+
+Santa was cheerful about this guide, especially that it contains the long avaited Christmas cookie statistics. 
+And perhaps the next task will be to extend this report with more Christmas cookies from around the world.
+
+We just scratched the surface of C<Excel::Writer::XLSX> with this short tutorial. You can find a comprehensive guide on CPAN, L<Excel::Writer::XLSX|https://metacpan.org/pod/Excel::Writer::XLSX>.
+This is a very useful and powerful module, which can deal with a range of tasks from the basics of generating reports to the heavy lifting of the work with VBA in Excel.
+To cover all of it's features would take a whole book.
```
Comment (openstrike) [inline 2022/incoming/create-your-excel-using-perl.pod:163]: s/it's/its/

File: 2022/incoming/create-your-excel-using-perl.pod [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+You can find the whole script on my Github L<cookies_test.pl|https://github.com/x1mandi/padvent_code/blob/main/cookies_test.pl>, so that You can test it on Your own and play with the module. \
```
Comment (openstrike) [inline 2022/incoming/create-your-excel-using-perl.pod:165]: The backslash at the end can be removed.

**Review by openstrike — CHANGES_REQUESTED** [review]
(Note - I've only reviewed `create-your-excel-using-perl.pod` here, not the other 2 articles.)

Nice article. I've suggested a large number of changes here but they are all minor - mostly typos and grammar.

In several places "You" or "Your" is capitalised in mid-sentence - these should all be lowercase. Really "I" is the only capital pronoun used in English these days.

The elf name Wunorse Openslae is inspired!

File: 2022/incoming/create-your-excel-using-perl.pod [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+=encoding utf8
+
+=head1 Create Your Excel using Perl!
+
+=head2 Intro
+
+A long-long time ago, during the Advent of 2001, there was an article about manipulating Excel spreadsheets using Perl, L<Spreadsheet::WriteExcel|https://perladvent.org/2001/14th/>.
+It's been a while now, so Santa decided to ask the Wise Old Elf to write another article about creating Excel files using Perl. 
+
+There were only 3 days until Advent, with lots of articles missing, so the Wise Old Elf immediately delegated this task to Wunorse Openslae, and he choose wisely - as always. Wunorse was a master of managing Excel files and started to work on that article immediately.
```
[the remark this answers is not in the document]
Comment (x1mandi) [reply 2022/incoming/create-your-excel-using-perl.pod:14]: Done.

File: 2022/incoming/create-your-excel-using-perl.pod [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+
+=head2 Preface
+
+Some historical background: The older module L<Spreadsheet::WriteExcel|https://metacpan.org/pod/Spreadsheet::WriteExcel> is for creating Excel files compatible with Excel 97, 2000, 2002, 2003 and 2007, mainly with C<.xls> extension. It is mostly compatible with Excel::Writer::XLSX, so L<migrating to the newer version|https://metacpan.org/pod/Spreadsheet::WriteExcel#Migrating-to-Excel::Writer::XLSX> should be as easy as a christmas pie. However such a migration is more like a heavy christmas pudding loaded with goodness and lots of issues might arise during the process.
```
[the remark this answers is not in the document]
Comment (x1mandi) [reply 2022/incoming/create-your-excel-using-perl.pod:18]: Done

File: 2022/incoming/create-your-excel-using-perl.pod [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+
+The old module comes along with C<Spreasheet::ParseExcel> just to name an example for reading Excel files. You can use C<Spreadsheet::ParseXLSX> or C<Spreadsheet::XLSX> to open and manipulate existing Excel files. This article deals with creating new Excel files.
```
[the remark this answers is not in the document]
Comment (x1mandi) [reply 2022/incoming/create-your-excel-using-perl.pod:20]: I have rewritten the whole sentence here, it was not really clear.

File: 2022/incoming/create-your-excel-using-perl.pod [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+=head2 Basics
+
+First things first: we need to import the module, initialze an xlsx file, add some worksheets, input some data on them and finally close the file, as we would do with a usual filehandle.
```
[the remark this answers is not in the document]
Comment (x1mandi) [reply 2022/incoming/create-your-excel-using-perl.pod:24]: Done

File: 2022/incoming/create-your-excel-using-perl.pod (hunk 2 above)
Comment (x1mandi) [reply 2022/incoming/create-your-excel-using-perl.pod:70]: Done

File: 2022/incoming/create-your-excel-using-perl.pod [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+	$format2->set_bold();
+	$format2->set_size( 12 );
+	$format2->set_bg_color( 'green' );
+	$format2->set_align('center');
+	$format2->set_border( 4 );
+
+	#Just some snowflakes as border
+	my $format3 = $merry->add_format( border => 4,);
+
+=end code
```
[the remark this answers is not in the document]
Comment (x1mandi) [reply 2022/incoming/create-your-excel-using-perl.pod:97]: Done

File: 2022/incoming/create-your-excel-using-perl.pod (hunk 3 above)
Comment (x1mandi) [reply 2022/incoming/create-your-excel-using-perl.pod:102]: Done

File: 2022/incoming/create-your-excel-using-perl.pod (hunk 4 above)
Comment (x1mandi) [reply 2022/incoming/create-your-excel-using-perl.pod:137]: done

File: 2022/incoming/create-your-excel-using-perl.pod (hunk 5 above)
Comment (x1mandi) [reply 2022/incoming/create-your-excel-using-perl.pod:152]: done

File: 2022/incoming/create-your-excel-using-perl.pod (hunk 6 above)
Comment (x1mandi) [reply 2022/incoming/create-your-excel-using-perl.pod:163]: done

File: 2022/incoming/create-your-excel-using-perl.pod (hunk 7 above)
Comment (x1mandi) [reply 2022/incoming/create-your-excel-using-perl.pod:165]: Done

**Review by x1mandi — COMMENTED** [review]
Committed all requested changes. Thank You, @openstrike. 

Please check!

**Comment by x1mandi** [comment]
> In several places "You" or "Your" is capitalised in mid-sentence - these should all be lowercase. Really "I" is the only capital pronoun used in English these days.
@openstrike 
Corrected this one as well.

File: 2022/incoming/create-your-excel-using-perl.pod [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+
+There were only 3 days until Advent, with lots of articles missing, so the Wise Old Elf immediately delegated this task to Wunorse Openslae, and he chose wisely - as always. Wunorse was a master of managing Excel files and started to work on that article immediately.
+
+=head2 Preface
+
+Some historical background: The older module L<Spreadsheet::WriteExcel|https://metacpan.org/pod/Spreadsheet::WriteExcel> is for creating Excel files compatible with Excel 97, 2000, 2002, 2003 and 2007, mainly with C<.xls> extension. It is mostly compatible with Excel::Writer::XLSX, so L<migrating to the newer version|https://metacpan.org/pod/Spreadsheet::WriteExcel#Migrating-to-Excel::Writer::XLSX> should be as easy as Christmas pie. However such a migration is more like a heavy Christmas pudding loaded with goodness and lots of issues might arise during the process.
+
+You can use C<Spreasheet::ParseExcel> for reading the older Excel files having C<.xls> extension, on the other hand you should use C<Spreadsheet::ParseXLSX> or C<Spreadsheet::XLSX> to open and manipulate existing Excel files. This article deals with creating Excel files in C<.xlsx> format.
```
Comment (openstrike) [inline 2022/incoming/create-your-excel-using-perl.pod:20]: This is indeed better.

One small change to make it even clearer - I presume you mean to use the newer modules on .xlsx files only so maybe add that?

eg: "on the other hand you should use C<Spreadsheet::ParseXLSX> or C<Spreadsheet::XLSX> to open and manipulate existing Excel files having the .xlsx extension"

File: 2022/incoming/create-your-excel-using-perl.pod (hunk 12 above)
[the remark this answers is not in the document]
Comment (openstrike) [reply 2022/incoming/create-your-excel-using-perl.pod:97]: Almost! Now it is just set_align which has missing spaces:

```
	$format2->set_color( 'red' );
	$format2->set_bold();
	$format2->set_size( 12 );
	$format2->set_bg_color( 'green' );
	$format2->set_align('center');
	$format2->set_border( 4 );```

File: 2022/incoming/create-your-excel-using-perl.pod [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+	$format2->set_border( 4 );
+
+	#Just some snowflakes as border
+	my $format3 = $merry->add_format( border => 4,);
+
+=end code
+
+=head2 Writing data into cells
+
+There are two ways to do that: You can use either the row-column notation or the A1 notation as described in L<Cell notation|https://metacpan.org/pod/Excel::Writer::XLSX#Cell-notation>.
```
Comment (openstrike) [inline 2022/incoming/create-your-excel-using-perl.pod:101]: s/You/you/

File: 2022/incoming/create-your-excel-using-perl.pod [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+        
+=end code
+
+Many things are happening here.
+
+First we get a list of countries sorted alphabetically.
+Second we are writing those countries into the first row, alongside some formats, defined above.
+Then comes the first usage of C<write()>, where we are using the A1 notation to start from the top left corner, then comes the reference to the C<@countries> array, followed by the format2.
+When we pass an array reference to C<write()>, it will write the elements of the array in the same row. It is equivalent to write_row().
+If You pass an array ref of array refs to C<write()>, it is the same as calling C<write_col()>, it will write to the columns, as demonstrated in the lines 22 and 27.
```
Comment (openstrike) [inline 2022/incoming/create-your-excel-using-perl.pod:152]: s/You/you/

**Review by openstrike — CHANGES_REQUESTED** [review]
Looks good - thanks. Just a few more minor changes and it should be ready to go.

**Review by x1mandi — COMMENTED** [review]
Next iteration of changes done.

**Comment by x1mandi** [comment]
> Looks good - thanks. Just a few more minor changes and it should be ready to go.

@openstrike Followed the suggestions, all done.

**Review by openstrike — APPROVED** [review]
Thanks - this article is ready to go now.

(again, for the mergers, I've only reviewed create-your-excel-using-perl.pod)

File: 2022/incoming/lapland-packaging-department-pt2.pod [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+
+=begin code 
+
+$ carton install --deployment cached
+
+=end code
+
+This will use the module tarballs from the C<vendor/> directory and create C<local/> with the installed modules.
+
+If no errors You may remove the C<vendor/> directory (optional).
```
Comment (oodler577) [inline 2022/incoming/lapland-packaging-department-pt2.pod:214]:
```suggestion
If there are no errors You may remove the C<vendor/> directory (optional).
```

File: 2022/incoming/lapland-packaging-department-pt2.pod [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+|-- ...  #Your stuff here
+|-- ...	 #Your stuff here
+`-- vendor
+	`-- cache
+	    |-- authors
+	    `-- modules
+
+=end code
+
+Now it's time to install the modules using Carton (which has been previously installed by Tonttu):
```
Comment (oodler577) [inline 2022/incoming/lapland-packaging-department-pt2.pod:204]:
```suggestion
Now it's time to install the modules using L<Carton> (which has been previously installed by Tonttu):
```

File: 2022/incoming/lapland-packaging-department-pt2.pod [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+
+=end code
+
+This will use the module tarballs from the C<vendor/> directory and create C<local/> with the installed modules.
+
+If no errors You may remove the C<vendor/> directory (optional).
+
+Santa reviewed the proof of concept and was proud of Tonttu for doing such a great job. 
+
+After sharing the how-to all the elves started to use the Carton packaging module and their job got a lot easier.
```
Comment (oodler577) [inline 2022/incoming/lapland-packaging-department-pt2.pod:218]:
```suggestion
After sharing the I<how-to> with them, all of the elves started to use the L<Carton> packaging module and their job got a lot easier.
```

**Review by oodler577 — CHANGES_REQUESTED** [review]
I will definitely be looking at Carton for some projects, this was a nice and accessible introduction to it!

**Comment by oodler577** [comment]
> Thanks - this article is ready to go now.
> 
> (again, for the mergers, I've only reviewed create-your-excel-using-perl.pod)

I took a pass on the other 2. Hope it helps!

**Comment by x1mandi** [comment]
@oodler577 Thank You for taking a look at the other 2 articles. I have committed Your suggestions, except the one about cabox, I don't consider it as an issue, it is a generic user of a cloud IDE, same for every user of the service.

**Comment by oalders** [comment]
Thanks, @x1mandi!

**Pull request merged**